core/arch/arm/plat-k3/drivers/ti_sci.c

```diff
@@ -213,6 +213,105 @@ int ti_sci_device_put(uint32_t id)
 	return ti_sci_device_set_state(id, 0, MSG_DEVICE_SW_STATE_AUTO_OFF);
 }
 
+int ti_sci_set_fwl_region(uint16_t fwl_id, uint16_t region,
+			  uint32_t n_permission_regs, uint32_t control,
+			  uint32_t permissions[FWL_MAX_PRIVID_SLOTS],
+			  uint64_t start_address, uint64_t end_address)
+{
+	struct ti_sci_msg_req_fwl_set_firewall_region req = { };
+	struct ti_sci_msg_resp_fwl_set_firewall_region resp = { };
+	struct ti_sci_xfer xfer = { };
+	int ret = 0;
+
+	ret = ti_sci_setup_xfer(TI_SCI_MSG_FWL_SET, 0,
+				&req, sizeof(req),
+				&resp, sizeof(resp),
+				&xfer);
+	if (ret)
+		return ret;
+
+	req.fwl_id = fwl_id;
+	req.region = region;
+	req.n_permission_regs = n_permission_regs;
+	req.control = control;
+	req.permissions[0] = permissions[0];
+	req.permissions[1] = permissions[1];
+	req.permissions[2] = permissions[2];
+	req.start_address = start_address;
+	req.end_address = end_address;
+
+	ret = ti_sci_do_xfer(&xfer);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+int ti_sci_get_fwl_region(uint16_t fwl_id, uint16_t region,
+			  uint32_t n_permission_regs, uint32_t *control,
+			  uint32_t (*permissions)[FWL_MAX_PRIVID_SLOTS],
```

Comment:
I'm not sure what this means, but it looks like `uint32_t permissions[FWL_MAX_PRIVID_SLOTS]` should be enough and avoid the funny looking assignments below.

Reply:
This is the first time I've seen an array passed in such a way (as a symbolic gesture) so I very much doubt that this is something people in general would expect for an output array.

A simple array would be preferable. Input only arrays are normally `const` while output arrays are without that attribute.

Reply:
It makes it easier for the user of the function too, they just pass all the output variables prefixed with `&`, even the array `permissions`:

```
ti_sci_get_fwl_region(fwl_id, sa2ul_region, 1,
					    &control, &permissions,
					    &start_address, &end_address);
```

Otherwise it would be the only output that doesn't have `&`.

I'm not inflexible here, if we want const and pass only the array itself then I can switch.

Reply:
Please use const instead.